napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,138 @@
+        _layers_to_task : dict
+            task storage for cancellation logic
+        _lock_layers_to_task : threading.RLock
+            lock to guard against changes to `_layers_to_task` when finding,
+            adding, or removing tasks.
+        """
+        self.events = EmitterGroup(source=self, ready=Event)
+        self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._layers_to_task: Dict[Tuple[Layer], Future] = {}
+        self._lock_layers_to_task = RLock()
+
+    def slice_layers_async(
+        self, layers: Iterable[Layer], dims: Dims
+    ) -> Future[dict]:
+        """This should only be called from the main thread.
+
+        Creates a new task and adds it to the _layers_to_task dict. Cancels
+        all tasks currently pending for that layer tuple.
+
+        Submitting multiple layers at one generates multiple requests (stored in a dict),
+        but only ONE task.
+
+        A slice submitted with multiple layers will only cancel if the same tuple of layers
+        is resubmitted. If a second slice is requested with only one layer from the first
+        slice, it will not be cancelled.
+        """
+        # Cancel any tasks that are slicing a subset of the layers
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        if existing_task := self._find_existing_task(layers):
+            logger.debug('Cancelling task for %s', layers)
+            existing_task.cancel()
+
+        # Not all layer types will initially be asynchronously sliceable.
+        # The following logic gives us a way to handle those in the short
+        # term as we develop, and also in the long term if there are cases
+        # when we want to perform sync slicing anyway.
+        requests = {}
+        for layer in layers:
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # create task for slicing of each request/layer
+        task = self._executor.submit(self._slice_layers, requests)
+
+        # store task for cancellation logic
+        self._layers_to_task[tuple(requests.keys())] = task
+
+        task.add_done_callback(self._on_slice_done)
+
+        return task
+
+    def shutdown(self) -> None:
+        """This should be called from the main thread when this is no longer needed."""
+        self._executor.shutdown()
+
+    def _slice_layers(self, requests: Dict) -> Dict:
+        """This can be called from the main or slicing thread.
+        Iterates through a dictionary of request objects and call the slice
+        on each individual layer."""
+        return {
+            layer: layer._get_slice(request)
+            for layer, request in requests.items()
+        }
+
+    def _on_slice_done(self, task: Future[Dict]) -> None:
```

Comment:
This can be called from either the main or slicing thread. We don't want users to directly call this, which is why it's private. Instead just describes what should happen when a task submitted by the executor finishes.

I can't think of a use case for calling this as a client. The intended public way to react to a slice being done is to connect to the `ready` event.